README.md
GNU MPC
=======

This gem provides Ruby bindings to the [GNU MPC](http://www.multiprecision.org/) library.

Methods
=======

"Initialization" Methods
------------------------

```ruby
MPC.new(7)      # (7,0)
MPC.new(7, 32)  # (7,0) with precision 32
MPC#prec        # precision if real and imag precisions are the same, 0 otherwise
MPC#prec2       # real and imaginary precisions, as a 2-element Array
MPC#prec=       # set precision
MPC#set_prec(precision, rounding_mode) # set precision
```

Projection and Decomposing Methods
----------------------------------

```ruby
MPC#real   # real part
MPC#imag   # imaginary part
MPC#arg    # argument
MPC#proj   # projection
```

Basic Arithmetic Methods
------------------------

```ruby
MPC#add(w)    # add with w
MPC#sub(w)    # difference with w
MPC#mul(w)    # multiply with w
MPC#div(w)    # divide by w
MPC#fma(w,x)  # multiple by w, then add with x
```

Power and Logarithm Methods
---------------------------

```ruby
MPC#sqrt      # square root
MPC#pow(w)    # raise to the power w
MPC#** w      # raise to the power w
MPC#exp       # exponential
MPC#log       # natural logarithm
MPC#log10     # base-10 logarithm
```

Trigonometric Methods
---------------------

The GNU MPC gem provides bindings to the following functions from MPC:

```ruby
MPC#sin    # sine
MPC#cos    # cosine
MPC#tan    # tangent
MPC#sinh   # hyperbolic sine
MPC#cosh   # hyperbolic cosine
MPC#tanh   # hyperbolic tangent
MPC#asin   # inverse sine
MPC#acos   # inverse cosine
MPC#atan   # inverse tangent
```

Miscellaneous Functions
-----------------------

```ruby
GMP::RandState#mpc_urandom  # uniformly distributed random complex number within
                            # [0,1] x [0,1]
```

...

Each of these methods accepts optional arguments to specify the rounding mode,
and precision, in the following fashions:

```
z = MPC.new(0, 1)
z.sin()                                    # default rounding mode; precision of the receiver is
                                           # applied to the return value
z.sin(MPC::MPC_RNDZZ)                      # MPC_RNDZZ rounding mode; precision of the receiver
                                           # is applied to the return value
z.sin(MPC::MPC_RNDNN, 64)                  # MPC_RNDNN rounding mode; precision of both real and
                                           # imaginary parts of return value is 64
z.sin(MPC::MPC_RNDNN, 64, 128)             # MPC_RNDNN rounding mode; precision
                                           # of real part of return value is 64,
                                           # imaginary part is 128
z.sin(:rounding => MPC::MPC_RNDZZ)         # MPC_RNDZZ rounding mode; precision
                                           # of the receiver is applied to the
                                           # return value
z.sin(:prec => 64)                         # default rounding mode; precision of
                                           # both real and imaginary parts of
                                           # return value is 64
z.sin(:pre_real => 64, :prec_imag => 128)  # default rounding mode; precision of
                                           # real part of return value is 64,
                                           # imaginary part is 128
```

Either the ordered list of arguments, or the options Hash may be passed; they
cannot be mixed.

Manual
======

There is also a PDF [manual](manual.pdf) in addition to this readme.


CHANGELOG
### 0.9.1

* added MPC#prec=, with documentation, and tests
* added MPC#set_prec, with documentation, and tests
* fixed and added tons of rdocs... tons

### 0.9.0

* documented MPC#proj
* added MPC#arg, with documentation, and tests
* added MPC#fma, wtih documentation, and tests
* added MPC#prec2, with tests
* added GMP::RandState#mpc_urandom, with tests
* added MPC#** and MPC#pow, with tests
* MPC#real and MPC#imag now take an optional precision, for the result.
* added tests for `#new` args, and fixed `#new(value, real_prec, imag_prec)` bug
* MPC#sub now accepts hash arguments (other bin-ops soon!)
* Bumped rspec and converting tests to just use expect().
* Travis is good.
* For Travis, whose VMs can only reach MPC 0.9-4, we know have support for MPC 0.9

### 0.8.2

* Added MPC#mul_i with tests

### 0.8.1

* Added MPC#/, MPC#div
* Added MPC#log10

### 0.8.0

* Most things


spec/add_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')

# All tests adapted from MPC 1.0.1's tests/add.dat
describe MPC, '#add' do
  it 'calculates the sum of two real MPCs' do
    op1 = MPC.new([1, 0])
    op2 = MPC.new([GMP::F("0x10000000000001p-105", 53, 16), GMP::F(0)])
    actual = op1.add(op2, MPC::MPC_RNDNN)
    expect(actual.real).to eq GMP::F("0x10000000000001p-52", 53, 16)
    expect(actual.imag).to eq GMP::F(0)

    actual = op1.add(op2, MPC::MPC_RNDZZ)
    expect(actual.real).to eq GMP::F("0x10000000000000p-52", 53, 16)
    expect(actual.imag).to eq GMP::F(0)

    actual = op1.add(op2, MPC::MPC_RNDUU)
    expect(actual.real).to eq GMP::F("0x10000000000001p-52", 53, 16)
    expect(actual.imag).to eq GMP::F(0)

    actual = op1.add(op2, MPC::MPC_RNDDD)
    expect(actual.real).to eq GMP::F("0x10000000000000p-52", 53, 16)
    expect(actual.imag).to eq GMP::F(0)
  end

  it 'calculates the sum of two real MPCs' do
    op2 = MPC.new([GMP::F(0), GMP::F("0x10000000000001p-105", 53, 16)])
    op1 = MPC.new([0, 1])
    actual = op1.add(op2, MPC::MPC_RNDNN)
    expect(actual.real).to eq GMP::F(0)
    expect(actual.imag).to eq GMP::F("0x10000000000001p-52", 53, 16)
  end
end


spec/arg_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')

describe MPC, "#arg" do
  it "returns the argument of an MPC generated with a Fixnum" do
    expect(MPC.new(0).arg).to eq GMP::F(0)
    expect(MPC.new(1).arg).to eq GMP::F(0)
    expect(MPC.new([0, 1]).arg).to eq((2*Math::PI) / 4)
    expect(MPC.new([1, 1]).arg).to eq((2*Math::PI) / 8)
  end

  it "returns the argument an MPC generated with a Float" do
    expect(MPC.new([1.0, Math.sqrt(3)]).arg).to eq((2*Math::PI) / 6)
  end
end


spec/cos_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')

# All tests adapted from MPC 1.0.1's tests/cos.dat
describe MPC, '#cos' do
  it 'calculates the cosine of a pure real argument' do
    data = [
      ["0x8a51407da8344p-52",   MPC.new([GMP::F(-1, 7), GMP::F(0, 7)])],
      ["0x8a51407da8344p-52",   MPC.new([GMP::F( 1, 7), GMP::F(0, 7)])]
    ]
    data.each do |expected, input|
      actual = input.cos(MPC::MPC_RNDNN, 50, 50)
      expect(actual.real).to eq GMP::F.new(expected, 50, 16)
      expect(actual.imag).to eq GMP::F(0)
    end
  end

  it 'calculates the cosine of a pure imaginary argument' do
    data = [
      ["0x18b07551d9f55p-48",   MPC.new([GMP::F(0, 7), GMP::F(-1, 7)])],
      ["0x18b07551d9f55p-48",   MPC.new([GMP::F(0, 7), GMP::F( 1, 7)])]
    ]
    data.each do |expected, input|
      actual = input.cos(MPC::MPC_RNDNN, 50, 50)
      expect(actual.real).to eq GMP::F.new(expected, 50, 16)
      expect(actual.imag).to eq GMP::F(0)
    end
  end

  it 'calculates the cosine of an argument with +1 and -1' do
    data = [
      ["0xd56f54b7a1accp-52", "-0xfd28666957478p-52",  MPC.new([GMP::F(-1, 7), GMP::F(-1, 7)])],
      ["0xd56f54b7a1accp-52",  "0xfd28666957478p-52",  MPC.new([GMP::F(-1, 7), GMP::F( 1, 7)])],
      ["0xd56f54b7a1accp-52",  "0xfd28666957478p-52",  MPC.new([GMP::F( 1, 7), GMP::F(-1, 7)])],
      ["0xd56f54b7a1accp-52", "-0xfd28666957478p-52",  MPC.new([GMP::F( 1, 7), GMP::F( 1, 7)])]
    ]
    data.each do |expected_real, expected_imag, input|
      actual = input.cos(MPC::MPC_RNDNN, 50, 50)
      expect(actual.real).to eq GMP::F.new(expected_real, 50, 16)
      expect(actual.imag).to eq GMP::F.new(expected_imag, 50, 16)
    end
  end

  it 'calculates the cosine of ieee-754 double precision numbers' do
    data = [
      [[514],                            [0], [0], ["0x1BBDD1808C59A3p-50", 53, 16], MPC::MPC_RNDNN],
      [["0x100FFFFFFFFFFFp-43", 53, 16], [0], [0], ["0x1BBDD1808C59A3p-50", 53, 16], MPC::MPC_RNDDD],
    ]
    data.each do |expected_real, expected_imag, input_real, input_imag, rounding_mode|
      actual = MPC.new([GMP::F.new(*input_real), GMP::F.new(*input_imag)]).cos(rounding_mode)
      expect(actual.real).to eq GMP::F.new(*expected_real)
      expect(actual.imag).to eq GMP::F.new(*expected_imag)
    end
  end
end


lib/mpc.rb
# Thank you, Nokogiri

require 'rbconfig'

require 'rubygems'
require 'gmp'

ENV['PATH'] = [File.expand_path(
  File.join(File.dirname(__FILE__), "..", "ext")
), ENV['PATH']].compact.join(';') if RbConfig::CONFIG['host_os'] =~ /(mswin|mingw|mingw32)/i

require File.dirname(__FILE__) + '/../ext/mpc'

class MPC
  def self.sprintf(format, *args)
    first_pct = format.index '%'
    result = format[0...first_pct]
    format.gsub(/(?<!%)%[0#+ ']*[0-9]*.?[0-9]*[a-zA-Z][^%]*/) do |fragment|
      arg = args.shift
      if fragment =~ /%[0#+ ']*[0-9]*.?[0-9]*[ZQF]/
        result << sprintf2(fragment, arg)
      elsif fragment =~ /%[0#+ ']*[0-9]*.?[0-9]*[PR]/ && GMP.const_defined?(:MPFR_VERSION)
        result << GMP::F.sprintf2(fragment, arg)
      else
        result << (fragment % arg)
      end
    end
    result
  end
end


ext/extconf.rb
#!/usr/bin/env ruby

require 'mkmf'

dir_config('mpc')

ok = true

unless have_header('gmp.h')
  $stderr.puts "can't find gmp.h, try --with-gmp-include=<path>"
  ok = false
end

unless have_library('gmp', '__gmpz_init')
  $stderr.puts "can't find -lgmp, try --with-gmp-lib=<path>"
  ok = false
end

unless (have_header('mpfr.h') and
    have_header('mpf2mpfr.h') and
    have_library('mpfr', 'mpfr_init'))
  ok = false
end
$CFLAGS += ' -DMPFR'

unless have_header('mpc.h')
  $stderr.puts "can't find mpc.h, try --with-mpc-include=<path>"
  ok = false
end

unless have_library('mpc', 'mpc_init2')
  $stderr.puts "can't find -lmpc, try --with-mpc-lib=<path>"
  ok = false
end

unless have_macro('SIZEOF_INTPTR_T')
  check_sizeof('intptr_t')
end

if try_compile('', '-O6')
  $CFLAGS += ' -Wall -W -O6 -g'
else
  $CFLAGS += ' -Wall -W -O3 -g'
end

if not ok
  raise "Unable to build, correct above errors and rerun"
end

create_makefile('mpc')


ext/ruby_mpc.h
#ifndef _RUBY_MPC_H_
#define _RUBY_MPC_H_

#include <stdio.h>
#include <ruby.h>
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <ruby_gmp.h>
#include <mpc.h>

#include <stdlib.h>

/*
 * (a comment borrowed from the Ruby GMP bindings:)
 * MP_???* is used because they don't have side-effects of single-element arrays mpc_t
 */
typedef __mpc_struct MP_COMPLEX;

#define mpc_get_struct(ruby_var,c_var) { Data_Get_Struct(ruby_var, MP_COMPLEX, c_var); }
#define mpc_make_struct(ruby_var,c_var) { ruby_var = Data_Make_Struct(cMPC, MP_COMPLEX, 0, r_mpc_free, c_var); }
#define mpc_make_struct_init(ruby_var,c_var) { mpc_make_struct(ruby_var,c_var); mpc_init (c_var); }
#define mpc_make_struct_init3(ruby_var,c_var,real_prec,imag_prec) { mpc_make_struct(ruby_var,c_var); mpc_init3 (c_var, real_prec, imag_prec); }
#define MPC_P(value) (rb_obj_is_instance_of(value,cMPC) == Qtrue)
#define mpc_set_bignum(var_mpc,var_bignum) {                         \
    VALUE tmp = rb_funcall (                                         \
        rb_funcall (var_bignum, rb_intern ("to_s"), 1, INT2FIX(32)), \
        rb_intern("upcase"),                                         \
        0);                                                          \
    mpc_set_str (var_mpc, StringValuePtr (tmp), 32);                 \
}

#define mpc_temp_alloc(var) { var=malloc(sizeof(MP_COMPLEX)); }
#define mpc_temp_init(var,prec) { mpc_temp_alloc(var); mpc_init2(var,prec); }
#define mpc_temp_free(var) { mpc_clear(var); free(var); }

#define r_mpc_set_z(var1, var2) (mpc_set_z (var1, var2, __gmp_default_rounding_mode))
#define r_mpc_set_d(var1, var2) (mpc_set_d (var1, var2, __gmp_default_rounding_mode))

/* EXPECTED_Cxxx macros */
#define EXPECTED_FXC "Expected GMP::F, Fixnum, or MPC"
#define EXPECTED_XBZFC "Expected Fixnum, Bignum, GMP::Z, Float, GMP::F, or MPC"

// MPC Rounding Modes
#define mpcrnd_get_struct(ruby_var,c_var) { Data_Get_Struct(ruby_var, int, c_var); }
#define mpcrnd_make_struct(ruby_var,c_var) { ruby_var = Data_Make_Struct(cMPC_Rnd, int, 0, 0, c_var); }
#define MPCRND_P(value) (rb_obj_is_instance_of(value, cMPC_Rnd) == Qtrue)
#define r_mpc_default_rounding_mode MPC_RNDNN

extern VALUE cMPC;
extern VALUE cMPC_Rnd;

extern void mpc_set_value(MP_COMPLEX *target, VALUE source, mpc_rnd_t rnd);
extern VALUE r_mpc_eq(VALUE self, VALUE arg);
extern VALUE r_mpc_cmp(VALUE self, VALUE arg);
extern int mpc_cmp_value(MP_COMPLEX *OP, VALUE arg);

extern void r_mpc_free(void *ptr);
extern mpc_rnd_t r_get_mpc_rounding_mode(VALUE rnd);
extern mpc_rnd_t r_get_default_mpc_rounding_mode();
extern mpc_rnd_t r_mpc_get_rounding_mode(VALUE rnd);
extern void init_mpcrnd();
extern void init_gmprandstate_mpc();

#endif /* _RUBY_MPC_H_ */


ext/gmprandstate.c
#include <ruby_mpc.h>

/**********************************************************************
 *    MPC Random Numbers                                              *
 **********************************************************************/

/*
 * call-seq:
 *   rand_state.mpc_urandom()
 *
 * From the MPC Manual:
 *
 * Generate a uniformly distributed random complex in the unit square [0,1] x [0,1].
 */
VALUE r_gmprandstate_mpc_urandom(int argc, VALUE *argv, VALUE self_val)
{
  MP_RANDSTATE *self;
  MP_COMPLEX *res;
  VALUE res_val;
  unsigned long prec = 0;

  if (argc > 1)
    rb_raise (rb_eArgError, "wrong # of arguments (%d for 0 or 1)", argc);

  mprandstate_get_struct (self_val, self);

  if (argc == 1) {
    if (FIXNUM_P (argv[0])) {
      if (FIX2INT (argv[0]) >= 2)
        prec = FIX2INT (argv[0]);
      else
        rb_raise (rb_eRangeError, "prec must be at least 2");
    } else {
      rb_raise (rb_eTypeError, "prec must be a Fixnum");
    }
  }

  mpc_make_struct (res_val, res);
  if (prec == 0) { mpc_init2 (res, mpfr_get_default_prec()); }
  else           { mpc_init2 (res, prec); }

  mpc_urandom (res, self);

  return res_val;
}

void init_gmprandstate_mpc()
{
  mGMP = rb_define_module("GMP");
  cGMP_RandState = rb_define_class_under(mGMP, "RandState", rb_cObject);

  // Complex Random Numbers
  rb_define_method(cGMP_RandState, "mpc_urandom", r_gmprandstate_mpc_urandom, -1);
}


ext/mpcrnd.c
#include <ruby_mpc.h>

VALUE r_mpcrnd_initialize(int argc, VALUE *argv, VALUE self);

VALUE r_mpcrndsg_new(int argc, VALUE *argv, VALUE klass)
{
  VALUE res;
  int *res_value;
  (void)klass;
  res_value = 0;

  mpcrnd_make_struct(res, res_value);
  rb_obj_call_init(res, argc, argv);
  return res;
}

/*
 * MPC_RNDNN 0000 0000
 * MPC_RNDNZ 0001 0000
 * MPC_RNDNU 0010 0000
 * MPC_RNDND 0011 0000
 * MPC_RNDNA ---- ----
 * MPC_RNDZN 0000 0001
 * MPC_RNDZZ 0001 0001
 * MPC_RNDZU 0010 0001
 * MPC_RNDZD 0011 0001
 * MPC_RNDZA ---- ----
 * MPC_RNDUN 0000 0010
 * MPC_RNDUZ 0001 0010
 * MPC_RNDUU 0010 0010
 * MPC_RNDUD 0011 0010
 * MPC_RNDUA ---- ----
 * MPC_RNDDN 0000 0011
 * MPC_RNDDZ 0001 0011
 * MPC_RNDDU 0010 0011
 * MPC_RNDDD 0011 0011
 * MPC_RNDDA ---- ----
 * MPC_RNDAN ---- ----
 * MPC_RNDAZ ---- ----
 * MPC_RNDAU ---- ----
 * MPC_RNDAD ---- ----
 * MPC_RNDAA ---- ----
 */
VALUE r_mpcrnd_initialize(int argc, VALUE *argv, VALUE self)
{
  VALUE mode, name, ieee754;
  char name_val[12];
  (void)argc;
  mode = argv[0];

  switch (FIX2INT(mode) % 16) {
  case 0:
    sprintf(name_val, "MPC_RNDN");
    ieee754 = rb_str_new2("(roundTiesToEven,");
    break;
  case 1:
    sprintf(name_val, "MPC_RNDZ");
    ieee754 = rb_str_new2("(roundTowardZero,");
    break;
  case 2:
    sprintf(name_val, "MPC_RNDU");
    ieee754 = rb_str_new2("(roundTowardPositive,");
    break;
  case 3:
    sprintf(name_val, "MPC_RNDD");
    ieee754 = rb_str_new2("(roundTowardNegative,");
    break;
  default:
    sprintf(name_val, "MPC_RNDN");
    ieee754 = rb_str_new2("(roundTiesToEven,");
  }

  switch (FIX2INT(mode) >> 4) {
  case 0:
    sprintf(name_val, "%sN", name_val);
    rb_str_cat(ieee754, "roundTiesToEven)", 16);
    break;
  case 1:
    sprintf(name_val, "%sZ", name_val);
    rb_str_cat(ieee754, "roundTowardZero)", 16);
    break;
  case 2:
    sprintf(name_val, "%sU", name_val);
    rb_str_cat(ieee754, "roundTowardPositive)", 20);
    break;
  case 3:
    sprintf(name_val, "%sD", name_val);
    rb_str_cat(ieee754, "roundTowardNegative)", 20);
    break;
  default:
    sprintf(name_val, "%sN", name_val);
    rb_str_cat(ieee754, "roundTiesToEven)", 16);
  }

  name = rb_str_new2(name_val);
  rb_iv_set (self, "@mode", mode);
  rb_iv_set (self, "@name", name);
  rb_iv_set (self, "@ieee754", ieee754);
  return Qnil;
}

VALUE r_mpcrnd_inspect(VALUE self)
{
  return rb_iv_get (self, "@name");
}

mpc_rnd_t r_get_default_mpc_rounding_mode()
{
  return r_get_mpc_rounding_mode (rb_const_get (cMPC, rb_intern ("MPC_RNDNN")));
}

mpc_rnd_t r_mpc_get_rounding_mode(VALUE rnd)
{
  VALUE mode;

  if (MPCRND_P(rnd)) {
    mode = rb_funcall (rnd, rb_intern("mode"), 0);
    if (FIX2INT(mode) < 0 || FIX2INT(mode) > 51) {
      rb_raise(rb_eRangeError, "rounding mode must be one of the rounding mode constants.");
    }
  } else {
    rb_raise(rb_eTypeError, "rounding mode must be one of the rounding mode constants.");
  }

  switch (FIX2INT (mode)) {
    case 0:
      return MPC_RNDNN;
    case 1:
      return MPC_RNDZN;
    case 2:
      return MPC_RNDUN;
    case 3:
      return MPC_RNDDN;
    case 16:
      return MPC_RNDNZ;
    case 17:
      return MPC_RNDZZ;
    case 18:
      return MPC_RNDUZ;
    case 19:
      return MPC_RNDDZ;
    case 32:
      return MPC_RNDNU;
    case 33:
      return MPC_RNDZU;
    case 34:
      return MPC_RNDUU;
    case 35:
      return MPC_RNDDU;
    case 48:
      return MPC_RNDND;
    case 49:
      return MPC_RNDZD;
    case 50:
      return MPC_RNDUD;
    case 51:
      return MPC_RNDDD;
    default:
      return MPC_RNDNN;
  }
}

void init_mpcrnd()
{
  ID new_id = rb_intern ("new");
  cMPC = rb_define_class ("MPC", rb_cNumeric);

  cMPC_Rnd = rb_define_class_under (cMPC, "Rnd", rb_cObject);

  rb_define_singleton_method (cMPC_Rnd, "new", r_mpcrndsg_new, -1);
  rb_define_method (cMPC_Rnd, "initialize", r_mpcrnd_initialize, -1);
  rb_define_method (cMPC_Rnd, "inspect", r_mpcrnd_inspect, 0);

  rb_define_attr (cMPC_Rnd, "mode",    1, 0);
  rb_define_attr (cMPC_Rnd, "name",    1, 0);
  rb_define_attr (cMPC_Rnd, "ieee754", 1, 0);

  rb_define_const(cMPC, "MPC_RNDNN", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(0)));
  rb_define_const(cMPC, "MPC_RNDNZ", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(16)));
  rb_define_const(cMPC, "MPC_RNDNU", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(32)));
  rb_define_const(cMPC, "MPC_RNDND", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(48)));
  rb_define_const(cMPC, "MPC_RNDZN", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(1)));
  rb_define_const(cMPC, "MPC_RNDZZ", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(17)));
  rb_define_const(cMPC, "MPC_RNDZU", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(33)));
  rb_define_const(cMPC, "MPC_RNDZD", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(49)));
  rb_define_const(cMPC, "MPC_RNDUN", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(2)));
  rb_define_const(cMPC, "MPC_RNDUZ", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(18)));
  rb_define_const(cMPC, "MPC_RNDUU", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(34)));
  rb_define_const(cMPC, "MPC_RNDUD", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(50)));
  rb_define_const(cMPC, "MPC_RNDDN", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(3)));
  rb_define_const(cMPC, "MPC_RNDDZ", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(19)));
  rb_define_const(cMPC, "MPC_RNDDU", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(35)));
  rb_define_const(cMPC, "MPC_RNDDD", rb_funcall (cMPC_Rnd, new_id, 1, INT2FIX(51)));
}
